### ISSUE TITLE
[Azure] ACI kubectl exec suppot

### ISSUE THREAD

**Issue by nonokangwei** [virtual-kubelet/virtual-kubelet#106]
when will the aci to support the kubectl exec? in the readme it set status as a current limitation.

**Comment by robbiezhang** [comment]
It's on the top of ACI's task list, and you will probably be able to use the `exec` in ACI in a couple weeks. The work to support it in VK will start after the API is available.

**Comment by nonokangwei** [comment]
@robbiezhang thanks robbie, look forward to this capability.

**Comment by rbitia** [comment]
@robbiezhang if you have a proposed design for exec add it here so we can get the community more involved w/ helping us add core VK functionality

**Comment by erjadi** [comment]
In case anyone else is looking into it as well, please be aware that the documentation on https://docs.microsoft.com/en-us/rest/api/container-instances/startcontainer/launchexec is inaccurate

The fields in entity "terminalSize" are actually called "rows" and "cols", not "row" and "column".

**Comment by erjadi** [comment]
Hi everyone,

I've given this a first stab and I'd love to know if anyone here would be willing to give me a hand. Right now I've only implemented exec in my fork, since I think it's quite a way off from becoming a pull request. It's currently very limited:

- Only implemented ACI provider
- Only works from Kubernetes UI (not kubectl command line)
- Clean up of terminal routines is buggy, kubectl proxy or az aks show might show errors
- Terminal resizing is not supported by ACI exec (and therefore not by my implementation)

### Installation
For those that are just interested in taking it for a spin,  please follow the manual setup as described at https://github.com/virtual-kubelet/virtual-kubelet/tree/master/providers/azure#manual-set-up
Before running the ```helm install``` you will need to set the following variable:
```CHART_URL=https://github.com/erjadi/virtual-kubelet/raw/master/charts/$VK_RELEASE.tgz```

### How it works
My implementation is based on my limited knowledge of kubelet/kubernetes and a lot of trial and error. I'd love to hear from any of you if my understanding is wrong in any part:

- Client sends shell command to kubernetes api server
- Kubernetes sends exec request to kubelet which looks something like this:
```POST /exec/default/nginx-deployment-85f76c5689-6tvjd/nginx?command=bash&amp;error=1&amp;input=1&amp;output=1&amp;tty=1 HTTP/1.1 
Host: 10.244.5.120:10250 
Connection: Upgrade 
Content-Length: 0 
Upgrade: SPDY/3.1 
User-Agent: dashboard/v1.8.3 
X-Forwarded-For: 172.31.36.1 
X-Stream-Protocol-Version: v4.channel.k8s.io 
X-Stream-Protocol-Version: v3.channel.k8s.io 
X-Stream-Protocol-Version: v2.channel.k8s.io 
X-Stream-Protocol-Version: channel.k8s.io 
```
- I'm aware that there are plans in Kubernetes to start using Websockets instead of SPDY, but I'm not sure of the timelines. Currently my fork only supports v4 SPDY
- The ACI provider will send a POST request to the container instnace to start exec. The body contains a command (by default /bin/bash) and the terminal size, both encoded in JSON.
- Azure will return a websocket URI and a password.
- Kubelet opens the websocket and sends the password, and from that moment the websocket acts as a remote stdin/out pipe to the container instance.
- The kubelet then starts a pair of goroutines to forward data between kubernetes and the container instance.

### Improvements to be made
- A larger group of us will have to agree on how much of the logic needs to be in the VK code and how much needs to be in the provider. For example: Currently I'm sending the ACI websocket password inside VK code rather than provider code, this most likely has to change but I'm unsure what the best implementation would be.
- Some code has been copied from kubernetes/pkg/kubelet/server/remotecommand/httpstream.go. This should be properly referenced.
- On Kubernetes <-> Kubelet communications, only SPDY v4 is currently supported, need support for v1/v2/v3? and Websockets?
- Kubectl exec commandline **currently does not work**, and I don't know why :(
- Where should goroutines that proxy the terminal session be spawned? In VK code or in provider code?
- Error handling (everywhere)
- Test coverage (exec functions aren't tested)
- Goroutines need to be properly rewritten, using channels for callback and defer for cleanup 

Again, I'd love to get some feedback from you about any of the above, even if it's just to confirm whether my fork works for you. I've only tested this for on AKS (since I'm working on ACI).

**Comment by rbitia** [comment]
@sflxn

**Comment by erjadi** [comment]
Thanks guys! @sflxn currently I believe ACI only supports exec, not attach, which was my first motivation to try my hand at this.

Once we do agree on the implementation for exec, I believe we could already add the interfaces for attach on the vk side since they are almost functionally the same.

**Comment by sflxn** [comment]
```
A larger group of us will have to agree on how much of the logic needs to be in the VK code and how much needs to be in the provider. For example: Currently I'm sending the ACI websocket password inside VK code rather than provider code, this most likely has to change but I'm unsure what the best implementation would be.
```
I think the answer to this is that the VK just needs 3 streams (for stdin, stdout, stderr) from whatever runtime the provider provides.  How you get the 3 streams should be up to the provider.  In this case, you're passing the password to ACI so that's setup work and should be in the provider itself.

```
Where should goroutines that proxy the terminal session be spawned? In VK code or in provider code?
```
I'm not sure what you mean by this.  Are you referring to the terminal resizing?  We can follow the same pattern that the kubelet uses.  It reads a resize stream and then sends the size in a channel.  The exec doesn't actually start till the size is ready.  I'm not exactly sure why kubernetes is doing this instead of just sending over the terminal size.  I assume it is to allow kubectl to set up the streams before retrieving the terminal size?

Anyway, I think we can handle this the same way as the kubelet.  The VK can pass a resize channel to the provider.

I have some comments about your code.  I'll add the comments inline.

**Comment by erjadi** [comment]
Hi @sflxn , thanks a lot for the feedback, really appreciate it. A lot of points I've figured out since I posted, and like you said there's a bunch of provider specific actions that need to be moved out of the VK. This implementation was mostly a way to see if I could make ACI exec work, and that's why the VK code is not up to scratch (and why it's not a PR but a fork). I'll do a clean implementation given your feedback and create a proper PR.

As for the goroutines, I was referring to all of the streams not just the terminal resizing, but I think it's clear to me now they need to be spawned in the provider code since they are specific to each provider.

Just some clarifications / remarks, but overall I agree with most of your points. 

> The exec doesn't actually start till the terminal resize occurs, which waits on the channel for the data. I'm not exactly sure why kubernetes is doing this instead of just sending over the terminal size in the query string. I assume it is to allow kubectl to set up the streams before retrieving the terminal size?

Yes, this is how I am doing it as well. Things are a bit inconvenient for ACI though, since there is no resizing functionality, and I use the terminal size from the first resize in the stream to actually initiate the shell with ACI. As far as I know, kubectl from command line does not send resize events.

>It appears you copy and pasted some code, but I don't see where you handle the error stream. In fact, I don't see an error stream in your context object. If you're wondering what I'm referring to, kubernetes sets up 5 streams in all, including 1 for stderr and 1 for error propgation.

Yes, I haven't implemented the parts that I could not use in ACI, and I'm aware these should be present in the VK code. This is one of the reasons I haven't made this a pull request yet :) 

>You increased the stream count when tty is set in the rest call's query string. The kubelet's code checks if the protocol supports resizing. Protocol 3 and 4 supports it, but the first two does not. Just for safety, I think we should do the same. In other words if tty && protocol.supportsResizing...

Yes, in the pull request I need to support more than just v4 and this is a part of that.

Regarding kubectl exec not working:
> I suspect it's not working because you're not handling the resize properly. You're throwing away the data.

I'm not sure this is the issue, since in the Azure case there is nothing I can do with the resize data anyway: ACI currently does not have a terminal resize function. Even if I didn't handle the resize requests, the stdin/stdout should come through, right?

**Comment by sflxn** [comment]
Hi @erjadi, in regards to the size events, are you sure about this?  I believe I saw the kubectl code that creates a stream just for size.  On the kubelet side, there is a goroutine that reads the stream and signals the size channel.

**Comment by erjadi** [comment]
Hi @sflxn, I've done a clean implementation now and I'm happy to see that kubectl exec -it works.

I may have not expressed myself clearly: I did not handle the resize events **after** I've created the websocket to the backend (since there is no resize on ACI). So in that sense both exec from the UI and the command line should be the same (i.e. I wait for the first resize event and then open up the websocket to the container instance). I think something must've gone wrong waiting for that first event...?

I'll push my code later today so everyone can see, there's still a few small open points. However as per your suggestion it's now fully based on the standard Executor interface so I can rely on the standard kubernetes libraries as well, without any code duplication.

Two points I'd like to ask for you help again:
- How to best pass namespace through the standard interface? I really like the idea of being able to just use remotecommand.ServeExec so I'd rather not modify that....
- httpstream waits for a fixed number of streams depending on the flags in the request to apiserver (in,out,err,tty). However, [it starts counting from 1](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/remotecommand/httpstream.go#L165) which I don't really understand? Basically If I don't hardcode tty to false, it times out waiting for streams :(
